### ISSUE TITLE
More highlighting issues

### ISSUE THREAD

**Issue by raph-amiard** [AdaCore/ada_language_server#436]
![hl_problem_1](https://user-images.githubusercontent.com/478727/80600072-29a38d80-8a2c-11ea-9d69-436f97f12022.png)
![hl_problem_2](https://user-images.githubusercontent.com/478727/80600074-2a3c2400-8a2c-11ea-9385-cf183a40d3de.png)
![hl_problem_3](https://user-images.githubusercontent.com/478727/80600078-2ad4ba80-8a2c-11ea-9225-af1bcc85c239.png)
![hl_problem_4](https://user-images.githubusercontent.com/478727/80600080-2ad4ba80-8a2c-11ea-8d17-eaac183b9750.png)

The last one with the wrongly highlighted comments is the most concerning to me. If you want to reproduce, the file `lsp-ada_handlers.adb` is from the ALS repo.

Honestly at this stage I think we should either:

- Have some kind of dumb "token highlighting" fallback on master, and maybe have a `highlighting-dev` branch where it is deactivated so that remaining holes are still visible, as discussed in #412 
- Completely roll back the new highlighter in favor of the old dumb one, and have the new one on a branch.

This should be complemented by extensive testing one way or another.

**Comment by Entomy** [comment]
@gcluzel since you've been contributing as well I'd like you to weight in.

There's a few ways we can approach this, which I will cover. To address the last statement, @setton has been adding classifier testing which I will also be contributing to when I can, so that is already being worked on. Unfortunantly unit testing for extensions was an afterthought for VS Code, so we've been operating without until I found someone else worked out a mechanism to do so. That being said, there's actually three options for how we can handle this, two that you've mentioned, and other which I mentioned as a late state migration option.

## LSP based classifier

This is the ideal that everything would eventually be based on. It is however, the slowest to develop, as I understand it, with very little documentation. It is required to handle certain parts of Ada's grammar, because of context sensitivity within the language that is impossible to resolve only from syntax. There are TreeSitter based extensions doing this, so the mechanism has been worked out despite little documentation. It is entirely dependent on the state of the LSP to work. I do not beleive it is currently in an adequate state for this yet, however, especially as it crashes quite often and that will stop all classifications being done through the LSP. No matter what, this is the eventual mechanism that will have to be done to handle certain ambiguities.

## Fallback

This would certainly work. I'm not sure how to merge the branch into master for updates without having it remove the fallback mechanism each merge. That's more of an issue with me being unfamiliar with git, because I primarily use a different VCS. But that's not a significant issue.

## Roll back

This would also work, of course. The older mechanism was far less tolerant of line breaks, however, which causes a lot of highlighting breaks for common styles. If we go this route, instead of a true roll back, I would suggest quickly implementing a very dumb shotgun classifier with only some basic attempts at highlighting types and other grammar parts which rely on position within constructs to be highlighted correctly. This would originally cause significant regressions for types and attributes. If we go this route, gradual migration to an LSP based classifier should begin quickly to handle the shortcomings of this approach.

**Comment by Entomy** [comment]
As for the specific issues shown here:

1) Use directives needs to be allowed within an additional group, declarative item iirc. Subroutine calls should be allowed within expression as well.

2) Same as 1. Comment might be an issue but I'm unfamiliar with the theme.

3) With if statements, it grabs the entire block first, then breaks it up into sections based on the then-else or then-end spans, then gets the if-then span and classifies everything in that as an expression. It should probably get that last part first, so that `and then` isn't mistaken as the same `then`, because these are entirely different semantics.

4) Comments are broken because it's not classifying them at all in that context. We need to add comments to all `patterns` groups.

**Comment by setton** [comment]
I think the first thing we need to do is reach a state where the highlighting is reasonably good - even if it's less sophisticated - for users, while we evaluate our options. To do this without putting undue pressure on ourselves to fix the sophisticated engine, I propose to do the following:

 * Right now: we move the current `ada.tmLanguage.json` to a separate directory, for instance "full_classifier", and put back the previous one in production. 

 * In the coming weeks (or months, depending on our availability), we beef up our testsuite, this is something that we'll need in any case.

 * In parallel we monitor the situation on the LSP: it looks like they are converging on semantic highlighting (see recent messages in [this issue](https://github.com/microsoft/language-server-protocol/issues/18)). As you say that's the ideal in the long term.

**Comment by raph-amiard** [comment]
> This would certainly work. I'm not sure how to merge the branch into master for updates without having it remove the fallback mechanism each merge. That's more of an issue with me being unfamiliar with git, because I primarily use a different VCS. But that's not a significant issue.

I agree that's the issue to figure out @Entomy. Could you maybe propose a patch for the fallback, and @setton can figure out the git workflow and explain it ?

**Comment by Entomy** [comment]
I'll save the current work and put it somewhere. A branch or something. And put back a shotgun classifier in its place, like before. From there, we can focus on the test suite, instead of the classifier, and keep an eye on both the LSP implementation and protocol API capabilities to determine how and when to migrate. Luckily this migration can be gradual, as the two can coexist.

I'll have a PR ready in a few days.

**Comment by setton** [comment]
Sounds good - thanks everyone for the feedback! I'll take care of the PR. 

What I'll do is move the work in a separate directory rather than a branch. Not a very strong opinion but my arguments for preferring this approach are:
  - this makes it possible to have an automated test for both the "sophisticated" and the "shotgun" classifiers if we want. 
  - doing ad-hoc merges and tests, or switching between one and the other is easy enough if they are in side by side directories

**Comment by Entomy** [comment]
@reznikmm that's the ideal way to handle it anyways. 👍
